README.md
# Request

request是基于[request](https://www.npmjs.com/package/request)封装的http客户端请求工具.

[![build status](http://git.beibei.com.cn/node/request/badges/master/build.svg)](http://git.beibei.com.cn/node/request/commits/master)
[![coverage report](http://git.beibei.com.cn/node/request/badges/master/coverage.svg)](http://git.beibei.com.cn/node/request/commits/master)
[![node](https://img.shields.io/badge/node-8.5.0-brightgreen.svg)](https://img.shields.io/badge/node-8.5.0-brightgreen.svg)
[![version](http://npm.beibei.com.cn/badge/v/@node/request.svg)](http://npm.beibei.com.cn/package/@node/request)
_ _ _

+ [安装](#安装)
+ [特性](#特性)
+ [用法](#用法)
+ [文档](#文档)
    + [Request](#request)
        + [get](#get)
        + [post](#post)

+ [联系我们](#联系我们)

_ _ _

## 安装

> npm install @node/request --registry="http://10.2.223.167:7001" --save

## 特性

相较于原始的 `request` 库，`@node/request` 针对我们自己的使用场景做了如下优化：

- 非常方便的连接预发调试

    使用者可以通过在请求头的 `Cookie` 中加入 `API-BETA` 来非常方便的连接预发调试，具体用法参见下方的 `预发调试`。

- 超时自动重连三次

    针对线上出现的 `ETIMEDOUT` 和 `ESOCKETTIMEDOUT` 错误，`@node/request` 会尝试重连三次，从而大幅减少此类错误。

- 支持 HTTPDNS

    `@node/request` 会自动识别「阿里云抖动等原因导致的域名解析错误」，并使用 `HTTPDNS` 获取正确 `IP`。

- 域名解析兜底

    遇到域名解析服务商故障也不用担心，`@node/request` 会自动从 `hconf` 获取兜底 IP。

- 由公司内部团队维护，可定制性高

    `@node/request` 现在由大前端架构团队维护，所以针对业务中具有一定通用性的场景，我们都可以添加到该基础库中，避免重复劳动，提升开发效率。

## 用法

```javascript
const request = require('@node/request');

await request.get('http://m.beidian.com');

```

### 预发调试

```javascript
const request = require('@node/request');

await request.get('https://api.beibei.com/mroute.html?method=xxx', {
    method: 'xxx'
}, {
    headers: {
        Cookie: ctx.headers.cookie, // 保证 `cookie` 中包含 `API-BETA=xx` ，其中 `xx` 必须是一个数字
    },
});

```

### 默认选项

通过 defaults 方法返回一个设置了默认选项的请求对象，接下来使用该对象发送的请求都将拥有这些默认选项。

```javascript
const request = require('@node/request');
const newRequest = request.defaults({
    json: false
});

await newRequest.get('https://api.beibei.com/mroute.html?method=xxx', {
    method: 'xxx'
});

```

_ _ _


## 文档

### Request

### get

**static Promiase.<IncomingMessage> get(string url, [object query], [object, options])**

| 参数 | 类型 | 默认值 | 是否必须 | 说明 |
|--------|--------|--------|--------|--------|
| url       |  String   | -   | 是 | 请求地址  |
| query       |  Object   | -   | 否 | 请求参数  |
| options       |  Object   | -   | 否 | 可以其他请求参数,相见[https://www.npmjs.com/package/request#requestoptions-callback](https://www.npmjs.com/package/request#requestoptions-callback)  |

```javascript
await Request.get('http://test.api?id=1');
```
- - -

### post

**static Promiase.<IncomingMessage> post(string url, [object form], [object, options])**

| 参数 | 类型 | 默认值 | 是否必须 | 说明 |
|--------|--------|--------|--------|--------|
| url       |  String   | -   | 是 | 请求地址  |
| form       |  Object   | -   | 否 | 请求体参数  |
| options       |  Object   | -   | 否 | 可以其他请求参数,相见[https://www.npmjs.com/package/request#requestoptions-callback](https://www.npmjs.com/package/request#requestoptions-callback)  |


```javascript
await Request.post('http://test.api', {id: 1});
```

_ _ _

### put

**static Promiase.<IncomingMessage> put(string url, [object form], [object, options])**

| 参数 | 类型 | 默认值 | 是否必须 | 说明 |
|--------|--------|--------|--------|--------|
| url       |  String   | -   | 是 | 请求地址  |
| form       |  Object   | -   | 否 | 请求体参数  |
| options       |  Object   | -   | 否 | 可以其他请求参数,相见[https://www.npmjs.com/package/request#requestoptions-callback](https://www.npmjs.com/package/request#requestoptions-callback)  |


```javascript
await Request.put('http://test.api', {id: 1});
```

_ _ _

### patch

**static Promiase.<IncomingMessage> patch(string url, [object form], [object, options])**

| 参数 | 类型 | 默认值 | 是否必须 | 说明 |
|--------|--------|--------|--------|--------|
| url       |  String   | -   | 是 | 请求地址  |
| form       |  Object   | -   | 否 | 请求体参数  |
| options       |  Object   | -   | 否 | 可以其他请求参数,相见[https://www.npmjs.com/package/request#requestoptions-callback](https://www.npmjs.com/package/request#requestoptions-callback)  |


```javascript
await Request.patch('http://test.api', {id: 1});
```

_ _ _

### del

**static Promiase.<IncomingMessage> del(string url, [object query], [object, options])**

| 参数 | 类型 | 默认值 | 是否必须 | 说明 |
|--------|--------|--------|--------|--------|
| url       |  String   | -   | 是 | 请求地址  |
| query       |  Object   | -   | 否 | 请求参数  |
| options       |  Object   | -   | 否 | 可以其他请求参数,相见[https://www.npmjs.com/package/request#requestoptions-callback](https://www.npmjs.com/package/request#requestoptions-callback)  |

```javascript
await Request.del('http://test.api?id=1');
```
- - -



## 联系我们

- 联系人 : 陈泽韦(zewei.chen@beibei.com)
- Issue  : [http://git.beibei.com.cn/node/request/issues](http://git.beibei.com.cn/node/request/issues)


CHANGELOG.md
# Change Log

### 1.0.5

_黄力钧_

* 更新：请求库

### 1.0.4

_黄力钧_

* 更新：请求库


### 1.0.3

_黄力钧_

* 更新：请求库


### 1.0.2

_黄力钧_

* 更新：请求库

### 1.0.1

_黄力钧_

* 更新：请求库

### 1.0.0

_黄力钧_

* 更新：请求库

test/ClinetTest.js
const test = require('ava').test;
const request = require('../index');
const httpdns = require('../src/httpdns');
const http = require('http');
const port = 10087;

test.before.serial(() => {
    http.createServer((req, res) => {
        const echo = {
            headers: req.headers,
            url: req.url,
            method: req.method,
        };

        if (echo.url !== '/timeout') {
            res.end(JSON.stringify(echo));
        }
    }).listen(port);
});

test.serial('test use http dns', async (t) => {
    // test empty url
    try {
        await httpdns(null);
        return t.fail();
    } catch (ex) { }

    // test https
    let httpsResp = await request.handleHostPort('get', 'https://m.beibei.com/ncm/activ/activity/fans_welfare.html', {});
    t.is(!!httpsResp.body, true);

    // test http
    let httpResp = await request.handleHostPort('get', 'http://sapi.beidian.com/item/detail/v1-29370257.html', {});
    t.is(!!httpResp.body, true);

    // test http use cache
    let httpResp2 = await request.handleHostPort('get', 'http://sapi.beidian.com/item/detail/v1-29370257.html', {});
    t.is(!!httpResp2.body, true);

    // test invalid hostname https
    try {
        await request.handleHostPort('get', 'https://ms.beibei.com/ncm/activ/activity/fans_welfare.html', {});
        t.fail();
    } catch (ex) { }

    // test invalid hostname http
    try {
        await request.handleHostPort('get', 'http://ma.beibei.com/ncm/activ/activity/fans_welfare.html', {});
        t.fail();
    } catch (ex) { }

    return t.pass();
});

test.serial('test Client.get success', async (t) => {
    let resp = await request.get('http://127.0.0.1:10087/get');

    t.is(resp.body.headers.host, '127.0.0.1:10087');
    t.is(resp.body.headers['accept-encoding'], 'gzip, deflate');
    t.is(resp.body.headers.accept, 'application/json');
    t.is(resp.body.url, '/get');
    t.is(resp.body.method, 'GET');

    return t.pass();
});

test.serial('test Client.post success', async (t) => {
    let resp = await request.post('http://127.0.0.1:10087/post');

    t.is(resp.body.method, 'POST');

    return t.pass();
});

test.serial('test Client.put success', async (t) => {
    let resp = await request.put('http://127.0.0.1:10087/put');

    t.is(resp.body.method, 'PUT');

    return t.pass();
});

test.serial('test Client.patch success', async (t) => {
    let resp = await request.patch('http://127.0.0.1:10087/patch');

    t.is(resp.body.method, 'PATCH');

    return t.pass();
});

test.serial('test Client.del success', async (t) => {
    let resp = await request.del('http://127.0.0.1:10087/del');

    t.is(resp.body.method, 'DELETE');

    return t.pass();
});

test.serial('test Client.get error', async (t) => {
    let resp;
    try {
        resp = await request.get('http://127.0.0.1:10088/post');
    } catch (ex) {
        resp = ex;
    }

    t.is(resp.address, '127.0.0.1');
    t.is(resp.code, 'ECONNREFUSED');
    t.is(resp.errno, 'ECONNREFUSED');
    t.is(resp.port, 10088);
    t.is(resp.method, 'GET');
    t.is(resp.url, 'http://127.0.0.1:10088/post');

    return t.pass();
});

test.serial('test Client get request', async (t) => {
    let req = await request.getRequest();

    t.is(typeof req, 'function');

    return t.pass();
});

test.serial('test Client get without url', async (t) => {
    try {
        await request.get();
    } catch (ex) {
        t.is(ex.message, 'url不能为空');
    }

    return t.pass();
});

test.serial('test the condition of generating abr', (t) => {
    const options0 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {});
    const options1 = request.composeOptions('get', 'api-xretail.beicang.com');
    const options2 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {
        abr: true,
    });
    const options3 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {
        abr: true,
        qs: {
            id: 1,
        },
    });
    const options4 = request.composeOptions('get', 'http://api.beicang.com/mroute.html?id=1', {
        abr: true,
        qs: {
            name: 'xxx',
        },
    });
    const options5 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {
        abr: true,
        body: {
            name: 'xxx',
        },
    });
    const options6 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {
        abr: true,
        form: {
            name: 'xxx',
        },
    });
    const options7 = request.composeOptions('get', 'http://api.beicang.com/mroute.html?id=1', {
        abr: true,
        body: {
            name: 'xxx',
        },
    });

    const options8 = request.composeOptions('get', 'http://api.beicang.com/mroute.html', {
        abr: true,
        qs: {
            id: 1,
        },
        body: {
            name: 'xxx',
        },
    });

    const hasAbr0 = options0.url.indexOf('_abr_') > -1;
    const hasAbr1 = options1.url.indexOf('_abr_') > -1;
    const hasAbr2 = options2.url.indexOf('_abr_') > -1;
    const hasAbr3 = options3.url.indexOf('_abr_') > -1;
    const hasAbr4 = options4.url.indexOf('_abr_') > -1;
    const hasAbr5 = options5.url.indexOf('_abr_') > -1;
    const hasAbr6 = options6.url.indexOf('_abr_') > -1;
    const hasAbr7 = options7.url.indexOf('_abr_') > -1;
    const hasAbr8 = options8.url.indexOf('_abr_') > -1;

    t.is(hasAbr0, false);
    t.is(hasAbr1, false);
    t.is(hasAbr2, true);
    t.is(hasAbr3, true);
    t.is(hasAbr4, true);
    t.is(hasAbr5, true);
    t.is(hasAbr6, true);
    t.is(hasAbr7, true);
    t.is(hasAbr8, true);
});

test.serial('test generate right abr in composeOptions', (t) => {
    let obj0 = {
        type: 'GET',
        url: '/mroute.html',
        query: {
            a: 1,
            b: 2,
        },
        body: {
            c: 3,
            d: 4,
        },
        unixTime: '1519722776',
    };

    let obj1 = {
        type: 'GET',
        url: '/mroute.html',
        query: {
            a: 11,
            b: 22,
        },
        body: {},
        unixTime: '1519722777',
    };

    let obj2 = {
        type: 'POST',
        url: '/mroute.html',
        query: {
            a: 24,
            b: 33,
        },
        body: {
            c: 11,
            d: 44,
        },
        unixTime: '1519722786',
    };
});

test('should return x-beta cookie when set api-cookie', (t) => {
    const options = request.composeOptions('localhost', 'get', {
        headers: {
            Cookie: 'API-BETA=57',
        },
    });

    t.is(options.headers.Cookie, '; HX-BETA=57');
});

test('should return x-beta cookie when set api-cookie and other cookie', (t) => {
    const options = request.composeOptions('localhost', 'get', {
        headers: {
            Cookie: 'test=1;API-BETA=57',
        },
    });

    t.is(options.headers.Cookie, 'test=1;HX-BETA=57');
});

test('should return _airborne_channel qs when airjs biz_key is set', (t) => {
    const airjs = global.airjs;
    global.airjs = {
        projectConfig: {
            biz_key: 'beibei',
        },
    };
    const options = request.composeOptions('localhost', 'get', {});

    t.deepEqual(options.qs, {
        _airborne_channel: 'beibei',
    });
    global.airjs = airjs;
});

test.serial('test Client timeout', async (t) => {
    try {
        await request.get('http://127.0.0.1:10087/timeout', {
            timeout: 1000,
        });
    } catch (ex) {
        let options = JSON.parse(ex.options);

        t.is(ex.code, 'ESOCKETTIMEDOUT');
        t.is(options.timeoutRetry, 0);
    }

    return t.pass();
});

test.serial('test Client enotfound', async (t) => {
    global.airjs = {
        bconfGet(path) {
            return Promise.resolve('183.136.223.213');
        },
    };

    await request.get('https://api2.beibei.com/index.html', {}, {
        timeout: 1000,
    });

    await request.get('https://api2.beibei.com/index.html', {}, {
        timeout: 1000 * 10,
    });

    await request.get('https://api2.beibei.com/index.html', {}, {
        timeout: 1000 * 10,
        timeoutRetry: 1,
    });

    return t.pass();
});

test.serial('test Client defaults', async (t) => {
    let startTime = Date.now();

    try {
        let newRequest = request.defaults({
            timeout: 1000,
        });

        await newRequest.get('http://127.0.0.1:10087/timeout');
    } catch (ex) {
        t.is(Date.now() - startTime < 10000, true);
    }

    return t.pass();
});

test.serial('test get error name', (t) => {
    t.is(
        request.getErrorName(),
        undefined
    );
    t.is(
        request.getErrorName({
            code: 'ENOTFOUND',
            syscall: 'getaddrinfo',
        }),
        'dns'
    );
    t.is(
        request.getErrorName({
            code: 'ETIMEDOUT',
            syscall: 'connect',
            address: '61.130.9.130',
        }),
        'hostport'
    );
    t.is(
        request.getErrorName({
            code: 'ETIMEDOUT',
        }),
        'timeout'
    );
    t.is(
        request.getErrorName({
            code: 'ESOCKETTIMEDOUT',
        }),
        'timeout'
    );
});

let sleep = (time) => new Promise((resolve) => {
    setTimeout(() => resolve(), time);
});

test.serial('test use http dns ttl', async (t) => {
    // test ttl
    await sleep(10000);

    let httpsResp2 = await request.handleHostPort('get', 'https://m.beibei.com/ncm/activ/activity/fans_welfare.html', {});
    t.is(!!httpsResp2.body, true);
});



src/httpdns.js
const request = require('request');
const crypto = require('crypto');
const Url = require('url');

const algorithm = 'des-ecb';
const key = 'z7X5)e/X';
const iv = Buffer.from('');

let getEncodeText = (plaintext) => {
    let cipher = crypto.createCipheriv(algorithm, key, iv);
    let crypted = cipher.update(plaintext, 'utf8', 'binary');
    crypted += cipher.final('binary');

    return crypted;
};

let getDecodeText = (cryptedText) => {
    let decipher = crypto.createDecipheriv(algorithm, key, iv);
    let plaintext = decipher.update(cryptedText, 'binary', 'utf8');
    plaintext += decipher.final('utf8');

    return plaintext;
};

let ipCache = {};

let fetchProxyFromCache = (protocol, hostname) => {
    let host = `${protocol}${hostname}`;
    let ip = ipCache[host];

    return ip ? Promise.resolve(ip) : Promise.reject();
};

let saveProxyToCache = (protocol, hostname, ip, ttl) => {
    let host = `${protocol}${hostname}`;
    const percent = 0.75; // 官方最佳实践推荐 https://cloud.tencent.com/document/product/379/3523

    ipCache[host] = ip;

    setTimeout(() => {
        ipCache[host] = undefined;
    }, ttl * 1000 * percent);
};

let parseHttpDnsResult = (result) => {
    let [ips, ttl] = result && result.split && result.split(',');
    let ip = ips && ips.split && ips.split(';')[0];
    ttl = Number(ttl);

    if (!ip || isNaN(ttl)) {
        return undefined;
    }

    return {
        ip,
        ttl,
    };
};

let fetchFromRemote = (protocol, hostname, backup = 0) => new Promise((resolve, reject) => {
    let useHttps = /^https:/.test(protocol);
    let dn = getEncodeText(hostname);
    let httpIps = ['119.29.29.29', '119.28.28.28'];
    let httpIp = httpIps[backup];
    let dnsUrl;

    if (useHttps) {
        dnsUrl = `https://dns.qq.com/d?dn=${dn}&id=64&ttl=1`;
    } else if (httpIp) {
        dnsUrl = `http://${httpIp}/d?dn=${dn}&id=64&ttl=1`;
    } else {
        reject(
            new Error('httpdns 服务暂不可用')
        );
        return;
    }

    request({
        method: 'GET',
        url: dnsUrl,
        timeout: process.env.NODE_ENV === 'test' ? 30 * 1000 : 5 * 1000,
    }, (error, response, body) => {
        if (error) {
            if (useHttps) {
                return reject(error);
            }
            return resolve(fetchFromRemote(protocol, hostname, backup + 1));
        }

        let result = getDecodeText(body);
        let parsedResult = parseHttpDnsResult(result);

        if (!parsedResult) {
            return reject(
                new Error(JSON.stringify({
                    message: `couldn't parse httpdns result`,
                    encodeText: body,
                    decodeText: result,
                    protocol,
                    hostname,
                }))
            );
        }

        let {
            ip,
            ttl,
        } = parsedResult;
        let proxy = useHttps ? `https://${ip}:443` : `http://${ip}:80`;

        saveProxyToCache(protocol, hostname, proxy, ttl);

        return resolve(proxy);
    });
});

module.exports = (url) => {
    try {
        let urlObj = Url.parse(url);
        let {
            protocol,
            hostname,
        } = urlObj;

        return Promise.resolve()
            .then(() => fetchProxyFromCache(protocol, hostname))
            .catch(() => fetchFromRemote(protocol, hostname));
    } catch (ex) {
        return Promise.reject(ex);
    }
};


example/request.js
// +----------------------------------------------------------------------
// | Beibei request
// +----------------------------------------------------------------------
// | HomePage : http://git.beibei.com.cn/node/request
// +----------------------------------------------------------------------
// | Author: 应杲臻 <gaozhen.ying@beibei.com>
// +----------------------------------------------------------------------

// const Request = require('request');

// // Request.get('http://127.0.0.1:10086/test', function(error, response) {
// //     console.log(response.body);
// // });
//
// const form = {
//     hello: 'world',
//     nihao: 'haoa'
// };
//
// const options = {
//     method : 'get',
//     url: 'http://127.0.0.1:10086?asdas=aa',
//     // form: form,
//     headers: {
//         Cookie: 'alan=walker'
//     },
//     qs: form
// };
//
// // Request.post('http://127.0.0.1:10086/test', {form: form}, function(error, response) {
// //     console.log(response.body);
// // });
// Request(options, function(error, response, body) {
//     console.log(body)
// });

const Request = require('../index');

(async () => {
    const resp = await Request.get('http://127.0.0.1:10087');
    // eslint-disable-next-line
    console.log(resp.body);
})();


src/Client.js
// +----------------------------------------------------------------------
// | Beibei request
// +----------------------------------------------------------------------
// | HomePage : http://git.beibei.com.cn/node/request
// +----------------------------------------------------------------------
// | Author: 应杲臻 <gaozhen.ying@beibei.com>
// +----------------------------------------------------------------------

const request = require('request');
const extend = require('extend');
const httpdns = require('./httpdns');
const {URL} = require('url');
const qs = require('qs');

class Client {
    /**
     * @public
     * @param {String} url      请求行地址
     * @param {Object} query    请求行参数
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static get(url, query = {}, options = {}) {
        options.qs = query;

        return this.send(Client.METHOD.GET, url, options);
    }

    /**
     * @public
     * @param {String} url      请求行地址
     * @param {Object} form     请求体参数
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static post(url, form = {}, options = {}) {
        options.form = form;

        return this.send(Client.METHOD.POST, url, options);
    }

    /**
     * @public
     * @param {String} url      请求行地址
     * @param {Object} form     请求体参数
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static put(url, form = {}, options = {}) {
        options.form = form;

        return this.send(Client.METHOD.PUT, url, options);
    }

    /**
     * @public
     * @param {String} url      请求行地址
     * @param {Object} form     请求体参数
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static patch(url, form = {}, options = {}) {
        options.form = form;

        return this.send(Client.METHOD.PATCH, url, options);
    }

    /**
     * @public
     * @param {String} url      请求行地址
     * @param {Object} form     请求体参数
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static del(url, query = {}, options = {}) {
        options.qs = query;

        return this.send(Client.METHOD.DELETE, url, options);
    }

    /**
     * 获取包含了默认值的的请求对象
     *
     * @returns {request}
     */
    static defaults(options) {
        class Defaults extends Client { }

        Defaults.DEFAULT_OPTIONS = extend(true, {}, this.DEFAULT_OPTIONS, options);

        return Defaults;
    }

    /**
     * 获取request客户端
     *
     * @returns {request}
     */
    static getRequest() {
        return request;
    }

    /**
     * @private
     * @param {String} method   请求方法
     * @param {String} url      请求行地址
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static send(method, url, options = {}) {
        /**
         * @type {Object}
         */
        let requestOptions;

        options.timeoutRetry === undefined && (options.timeoutRetry = 3);

        return new Promise((resolve, reject) => {
            try {
                requestOptions = this.composeOptions(method, url, options);
            } catch (e) {
                return reject(e);
            }

            return request(requestOptions, (error, response, body) => {
                if (error) {
                    let errorName = this.getErrorName(error);
                    let promise;
                    let sendError = (err) => {
                        if (Object.prototype.toString.call(err) === '[object Error]') {
                            err.method = method;
                            err.url = url;
                            err.options = JSON.stringify(options);
                        }

                        return reject(err);
                    };

                    if (errorName === 'timeout' && options.timeoutRetry) {
                        promise = this.handleTimeout(method, url, options);
                    } else if (errorName === 'rst' && options.timeoutRetry) {
                        // use handleTimeout retry.
                        promise = this.handleTimeout(method, url, options);
                    } else if (errorName === 'dns' && global.airjs && typeof global.airjs.bconfGet === 'function') {
                        promise = this.handleDns(method, url, options);
                    } else if (errorName === 'hostport' && !options.proxy) {
                        promise = this.handleHostPort(method, url, options);
                    }

                    if (promise) {
                        return promise
                            .then((result) => resolve(result))
                            .catch((err) => sendError(err || error));
                    }

                    return sendError(error);
                }

                return resolve(response);
            });
        });
    }

    /**
     * @private
     * @param {String} method   请求方法
     * @param {String} url      请求行地址
     * @param {Object} options  请求选项
     * @return {Promise.<IncomingMessage>}
     */
    static composeOptions(method, url, options = {}) {
        if (!url) {
            throw new Error('url不能为空');
        }

        let canSupportAbr = /\/\/api\./.test(url) || /\/\/api-xretail\./.test(url);

        if (canSupportAbr && options.abr) {
            const _url = new URL(url);

            const _query = qs.parse(_url.search, {
                ignoreQueryPrefix: true,
            });

            let {
                qs: query = {}, body = {}, form = {},
            } = options;

            let joinChar = url.indexOf('?') > -1 ? '&' : '?';
            url += `${joinChar}`;
        }

        if (global.airjs && global.airjs.projectConfig && global.airjs.projectConfig.biz_key) {
            options.qs = {
                ...options.qs,
                _airborne_channel: global.airjs.projectConfig.biz_key,
            };
        }

        let cookie = options && options.headers && options.headers.Cookie;

        if (cookie && cookie.match) {
            let match = cookie.match(/API-BETA=(\d+)/);

            if (match && match[1]) {
                let xBeta = Number(match[1]);

                if (!isNaN(xBeta)) {
                    options.proxy = 'http://212.129.209.215';

                    cookie = cookie.replace(/(HX|API)-BETA=\d+;?\s?/g, '').trim();

                    if (cookie[cookie.length - 1] === ';') {
                        cookie += `HX-BETA=${xBeta}`;
                    } else {
                        cookie += `; HX-BETA=${xBeta}`;
                    }
                    options.headers.Cookie = cookie;
                }
            }
        }

        return extend(true, {}, this.DEFAULT_OPTIONS, {
            method,
            url,
        }, options);
    }

    static getErrorName(error = {}) {
        let {
            code,
            syscall,
            address,
        } = error;

        if (code === 'ENOTFOUND' && syscall === 'getaddrinfo') {
            return 'dns';
        }

        if (code === 'ETIMEDOUT' && syscall === 'connect' && address) {
            return 'hostport';
        }

        if (code === 'ECONNRESET' && syscall === 'read') {
            return 'rst';
        }

        if (code === 'ETIMEDOUT' || code === 'ESOCKETTIMEDOUT') {
            return 'timeout';
        }

        return undefined;
    }

    static handleTimeout(method, url, options = {}) {
        options.timeoutRetry -= 1;
        return this.send(method, url, options);
    }

    static handleDns(method, url, options = {}) {
        return global.airjs
            .bconfGet('/bconf/ops/monitor_dns/domainresolve_ip')
            .then((ip) => {
                let proxy;
                let useHttps = /^https:/.test(url);
                if (useHttps) {
                    proxy = `https://${ip}:443`;
                } else {
                    proxy = `http://${ip}:80`;
                }
                if (options.proxy === proxy) {
                    return Promise.reject();
                }
                url = url.replace(/^https?/, 'http');
                options.proxy = proxy;
                return this.send(method, url, options);
            });
    }

    static handleHostPort(method, url, options = {}) {
        return httpdns(url)
            .then((proxy) => {
                url = url.replace(/^https?/, 'http');
                options.proxy = proxy;

                return this.send(method, url, options);
            });
    }
}

Client.DEFAULT_OPTIONS = {
    timeout: 15 * 1000,
    json: true,
    gzip: true
};

Client.METHOD = {
    GET: 'GET',
    POST: 'POST',
    PUT: 'PUT',
    PATCH: 'PATCH',
    DELETE: 'DELETE',
};

module.exports = Client;
